File: typescript/maroctax/src/frais-professionnels.ts
import type { FiscalConfig } from './types.js';
import { roundMoney } from './utils.js';

/**
 * Professional expenses deduction (frais professionnels).
 * 20% of (gross − CNSS − AMO), capped monthly.
 * @param grossAfterSocial - Gross minus CNSS and AMO.
 * @param config - Fiscal configuration for the payroll year.
 */
export function calculateFraisProfessionnels(
  grossAfterSocial: number,
  config: FiscalConfig,
): number {
  const uncapped = grossAfterSocial * config.fraisProfessionnelsRate;
  return roundMoney(Math.min(uncapped, config.fraisProfessionnelsMonthlyCap));
}

File: dart/maroctax/lib/src/frais_professionnels.dart
import 'fiscal_config.dart';
import 'types.dart';
import 'utils.dart';

/// Professional expenses deduction (frais professionnels).
double calculateFraisProfessionnels(
  double grossAfterSocial, [
  FiscalConfig? config,
]) {
  final cfg = config ?? getFiscalConfig();
  final uncapped = grossAfterSocial * cfg.fraisProfessionnelsRate;
  return roundMoney(
    uncapped < cfg.fraisProfessionnelsMonthlyCap
        ? uncapped
        : cfg.fraisProfessionnelsMonthlyCap,
  );
}
